Issue #214 - MbArray warning

milosstojanovic opened the pull request:
It works in following cases:
`def foo[@miniboxed T: ClassTag] = {
  new Array[T](5)
}`
and 
`new Array[Int](5)`
but I have issues to warn in this case:
`def foo[@miniboxed T: ClassTag](t: T) = {
  Array[T](t, t, t, t)
}`

I'm not sure how to match `Array.apply` method. Can you help me to match it in `MiniboxCommitTreeTransformer`? I tried with 
`case tree@Apply(apply , List(_, evidence)) if apply.symbol == Array_apply =>` 
but
`apply.symbol == Array_apply` gives `false` and I think that `apply` is not specific enough?

Commit:
Issue #214 - MbArray warning

VladUreche commented:
Nice work Milos!

Why do you warn on `Array[Int]`? I think if it's not generic + with a miniboxed type parameter, we shouldn't warn at all, right?

Regarding the `Array[T](t,t,t,t)` here's the desugared version of the code is:

```
scala.Array.apply[T](scala.this.Predef.genericWrapArray[T](Array[T]{C.this.t, C.this.t, C.this.t, C.this.t}), C.this.evidence$1)
```

In this case, you have to match the [`ArrayModule_genericApply`](https://github.com/scala/scala/blob/2.11.x/src/reflect/scala/reflect/internal/Definitions.scala#L440) symbol instead `Array_apply`, which belongs to the `Array` class (and not the `Array` object, also called module internally).

If you warn about the element-by-element constructor `Array[T](...)`, we should also have one for `MbArray`: `MbArray.apply[T](t: T*)` which could be transformed to `MBArray.clone[T](...)`. Up to it?

Commit:
Issue #214 - Warning for Array.apply[T](t: T*); add MbArray.apply[T](t: T*)

milosstojanovic commented:
@VladUreche 
- Ok, removed warning for `new Array[Int]`.
- Added warning for `Array[T](t,t,t,t)` (I match it in InteropCommitTreeTransformer to have a position) and implementation of `MbArray.apply[T](t: T*)`

VladUreche commented on components/runtime/src/scala/MbArray.scala:
```diff
@@ -64,6 +64,14 @@ object MbArray {
+    for (x <- xs.iterator) { mb_array.update(i, x); i += 1 }
+    mb_array
+  }
+
```
Milos, the problem with this version of `apply` is that it will always create the generic version of the `MbArray`, specifically `MbArray_L`. To create specialized arrays, we have two options:
- either redirect `apply` to an optimized version, as we do for `empty` and `clone` ([this code](https://github.com/milosstojanovic/miniboxing-plugin/blob/issue214/components/plugin/src/miniboxing/plugin/transform/minibox/commit/MiniboxCommitTreeTransformer.scala#L256-L280) redirects `empty` and `clone` to [their optimized implementations](https://github.com/milosstojanovic/miniboxing-plugin/blob/issue214/components/runtime/src/miniboxing/runtime/array/MbArrayOpts.java#L140-L206))
- add the `@miniboxed` annotation to the `apply` method, which should specialize `MbArray.empty` to create the correct version of the array (I personally don't like this option, as it adds to the recursive dependencies, which are already quite heavy, but it's also much easier to implement)

Also, please add the following testcase:

``` scala
println(MbArray(100, 200, 300).getClass) // should print "MbArray_I"
println(MbArray(1.0, 2.0, 3.0).getClass) // should print "MbArray_D"
println(MbArray("1", "2", "3").getClass) // should print "MbArray_L"
```

VladUreche replied:
Also, please don't forget to add your name to the `MbArray.scala` authors list.

VladUreche commented on components/plugin/src/miniboxing/plugin/transform/interop/commit/InteropCommitTreeTransformer.scala:
```diff
@@ -159,6 +159,15 @@ trait InteropCommitTreeTransformer extends TypingTransformers {
+                minibox.suboptimalCodeWarning(tree0.pos, minibox.UseMbArrayInsteadOfArrayWarning.msg)
+            }
+            super.transform(tree0)
+
```
As far as I can tell, this doesn't cover the case of `new Array[T](123)`. Is that handled elsewhere?

milosstojanovic replied:
Yes, here: https://github.com/milosstojanovic/miniboxing-plugin/blob/issue214/components/plugin/src/miniboxing/plugin/transform/minibox/commit/MiniboxCommitTreeTransformer.scala#L167

VladUreche replied:
Ok, that's a bit messy. And there's no way to move this warning to `MiniboxCommitTreeTransformer`?

VladUreche replied:
Worst case scenario, it will warn 3 times, for each class variant. Which I think gets filtered out anyway.

milosstojanovic replied:
The problem if I put it there is that Position of symbol will always  be NoPosition and i couldn't show the warning so that is the reason I had to move it to `InteropCommitTreeTransformer`

VladUreche replied:
Okay, makes sense. I will change it once I fix #168.

VladUreche commented on components/plugin/src/miniboxing/plugin/transform/minibox/commit/MiniboxCommitTreeTransformer.scala:
```diff
@@ -252,24 +253,30 @@ trait MiniboxCommitTreeTransformer extends TypingTransformers {
               case Some(minibox.Boxed)                            => super.transform(tree)
               case None if tpe1 <:< AnyRefTpe                     => super.transform(tree)
-              case None if ScalaValueClasses.contains(typeSymbol) => specialize(PartialSpec.valueClassRepresentation(typeSymbol))
+              case None if ScalaValueClasses.contains(typeSymbol) => specialize(PartialSpec.valueClassRepresentation(typeSymbol), true)
```
What is the difference between these `specialize(repr, false)` and `specialize(repr, true)` cases? As far as I can tell, in both cases you'll get a `WrappedArray` with the boxed arguments (even if they are miniboxed, they will be transformed into the boxed representation and put into the WrappedArray)

milosstojanovic replied:
In the case: `specialize(repr, true)`, the element of the WrappedArray is of type Int and by doing `minibox2int(array(index))`, the cast exception is thrown. So, all I need in that case is to cast it to Int `using array(index).asInstanceOf[Int]`. When `primType` is false, elements are miniboxed (either of type Long or Double), so I  have to transform them using `minibox2smth()` and cannot just cast them to primitive type using `asInstanceOf`. That's why I have to handle those two cases separetly

VladUreche replied:
> So, all I need in that case is to cast it to Int using array(index).asInstanceOf[Int].

Wow, that might actually be a miniboxing bug...

VladUreche replied:
I just filed it as https://github.com/miniboxing/miniboxing-plugin/issues/230.

VladUreche replied:
Okay, in that case, let's keep the current design until I fix the bug.

milosstojanovic replied:
Ok

VladUreche commented on components/runtime/src/miniboxing/runtime/array/MbArrayImpls.scala:
```diff
@@ -14,6 +16,17 @@ class MbArray_L[T](private[array] final val array: Array[AnyRef]) extends MbArra
+      }
+      array2
+    })
+
```
This is a really nice approach!

VladUreche commented on components/runtime/src/miniboxing/runtime/array/MbArrayOpts.java:
```diff
@@ -204,4 +206,48 @@ else if (mbArray instanceof MbArray_V<?>)
+  @SuppressWarnings("unchecked")
+  public static final <T> MbArray<T> mbArray_apply_constr_prim_D(Object array, byte T_Tag) {
+	return mbArray_clone_D(((WrappedArray)array).array(), T_Tag);
+  }
```
Why do you have this distinction between primitives and miniboxed values?

VladUreche commented on tests/correctness/resources/miniboxing/tests/compile/gh-bug-214.scala:
```diff
@@ -0,0 +1,37 @@
+
+  def foo[@miniboxed T](t: T) = {
+    MbArray[T](t, t, t)
+  }
```
Can you please test what happens if we have:

```
  def foo2[T](t: T) = {
    MbArray[T](t,t,t,t) // should warn T should be miniboxed
    new MbArray[T](10)  // should warn again
  }
```

milosstojanovic replied:
```
scala> def foo[T](t: T) = MbArray[T](t, t, t, t)
<console>:10: warning: The following code instantiating an MbArray object cannot be optimized since the type argument is not a primitive type (like Int), a miniboxed type parameter or a subtype of AnyRef. This means that primitive types could end up boxed:
       def foo[T](t: T) = MbArray[T](t, t, t, t)
                          ^
foo: [T](t: T)MbArray[T]

scala> def foo[T](t: T) = new MbArray[T](10)
<console>:10: error: class MbArray is abstract; cannot be instantiated
       def foo[T](t: T) = new MbArray[T](10)
                          ^
```

VladUreche replied:
Sorry, that should have been:

```
  def foo2[T](t: T) = {
    MbArray[T](t,t,t,t) // should warn T should be miniboxed
    MbArray.empty[T](10)  // should warn again
  }
```

Can you please include them in the test?

milosstojanovic replied:
This is the output:

```
scala>  def foo2[T](t: T) = {
     |     MbArray[T](t,t,t,t) // should warn T should be miniboxed
     |     MbArray.empty[T](10)  // should warn again
     |   }
<console>:8: warning: The following code instantiating an `MbArray` object cannot be optimized since the type argument is not a primitive type (like Int), a miniboxed type parameter or a subtype of AnyRef. This means that primitive types could end up boxed:
           MbArray[T](t,t,t,t) // should warn T should be miniboxed
           ^
<console>:9: warning: The following code instantiating an `MbArray` object cannot be optimized since the type argument is not a primitive type (like Int), a miniboxed type parameter or a subtype of AnyRef. This means that primitive types could end up boxed:
           MbArray.empty[T](10)  // should warn again
                   ^
foo2: [T](t: T)MbArray[T]

scala> 

```

I'll include them in the test as well.

VladUreche commented on components/plugin/src/miniboxing/plugin/metadata/MiniboxMetadataWarnings.scala:
```diff
@@ -158,17 +158,20 @@ trait MiniboxMetadataWarnings {
+  object UseMbArrayInsteadOfArrayWarning {
+    def msg: String = "Use MbArray instead of Array and benefit from miniboxing specialization. " +
+                      "For more details about MbArrays, please check the following link: " +
+                      "http://scala-miniboxing.org/arrays.html"
```
How about:

```
   def msg: String = "Use MbArray instead of Array to eliminate the need for ClassTags and " +
                     "benefit from seamless interoperability with the miniboxing specialization. " +
                     "For more details about MbArrays, please check the following link: " +
                     "http://scala-miniboxing.org/arrays.html"
```

milosstojanovic replied:
:+1:

VladUreche commented:
@milosstojanovic, great work! I nitpicked on some things and there's but important thing I don't understand: why separate the case of primitives and other things? Either way, the wrapped array will be of type Array[T], so you can access element by element. I would only do:

```
  @SuppressWarnings("unchecked")
  public static final <T> MbArray<T> mbArray_apply_ctor_J(WrappedArray<T> array, byte T_Tag) {
    return mbArray_clone_J(array.array(), T_Tag);
  }
```

And then you don't need the `WrappedArray` constructors anymore :)

Or am I missing something?

Commit:
MbArray warning content

milosstojanovic commented:
So, this is the exception that is thrown when I handle both primitive and miniboxed elements in the same way as you suggested:

```
Caused by: java.lang.ClassCastException: java.lang.Long cannot be cast to java.lang.Integer
    at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:105)
    at miniboxing.runtime.array.MbArray_I.<init>(MbArrayImpls.scala:53)
    at miniboxing.runtime.array.MbArrayOpts.mbArray_clone_J(MbArrayOpts.java:182)
    at miniboxing.runtime.array.MbArrayOpts.mbArray_apply_constr_prim_J(MbArrayOpts.java:246)
    at Test$.foo$n$J(newSource1.scala:15)
    at Test$.main(newSource1.scala:26)
    at Test.main(newSource1.scala)
    ... 51 more


```

VladUreche commented:
Right, that's bug #230. :((

VladUreche commented:
When you address the additional 2 tests, LGTM! Thanks a lot @milosstojanovic!

Commit:
Issue #214 - test updated

milosstojanovic commented:
Just added them :)

VladUreche commented:
LGTM!
